File: Python/833. Find And Replace in String.py
class Solution:
    def findReplaceString(self, S: str, indexes: List[int], sources: List[str], targets: List[str]) -> str:
        taken = 0
        replace = dict()
        for i, idx in enumerate(indexes):
            if S[idx: idx + len(sources[i])] == sources[i]:
                for j in range(idx, idx + len(sources[i])):
                    taken |= 1 << j
                replace[idx] = targets[i]
        idx_set = set(indexes)
        new_s = ""
        for i in range(len(S)):
            if taken & (1 << i) == 0:
                new_s += S[i]
            elif i in idx_set:
                new_s += replace[i]
        return new_s

File: 2021/833. Find And Replace in String.cpp
class Solution {
public:
    string findReplaceString(string s, vector<int>& indices, vector<string>& sources, vector<string>& targets) {
        map<int, string> sourceMap, targetMap;
        for (int i = 0; i < indices.size(); i++) {
            sourceMap[indices[i]] = sources[i];
            targetMap[indices[i]] = targets[i];
        }
        string res;
        int i = 0;        
        while (i < s.length()) {
            if (sourceMap.find(i) != sourceMap.end() && s.substr(i, sourceMap[i].length()) == sourceMap[i]) {
                res += targetMap[i];
                i += sourceMap[i].length();
            } else {
                res += s[i];
                i++;
            }
        }
        return res;
    }
};